### PR TITLE
Auto-Infer `mappings` Argument for `SmoothQuantModifier` Based on Model Architecture

### PR DESCRIPTION
**Pull request by rahul-tuli**
### Description:
This PR introduces a feature that automatically infers the `mappings` argument for the `SmoothQuantModifier` based on the model architecture, eliminating the need for manual specification of layer mappings.

#### **Before:**
In the prior implementation, users had to manually define layer mappings, as shown below:
```yaml
quantization_stage:
  quantization_modifiers:
    SmoothQuantModifier:
      smoothing_strength: 0.5
      mappings: [
        [["re:.*q_proj", "re:.*k_proj", "re:.*v_proj"], "re:.*input_layernorm"],
        [["re:.*gate"], "re:.*post_attention_layernorm"]
      ]
      ignore: ["lm_head"]
```

#### **Now:**
With this update, the `SmoothQuantModifier` automatically infers the mappings based on the architecture, simplifying the configuration:
```yaml
quantization_stage:
  quantization_modifiers:
    SmoothQuantModifier:
      smoothing_strength: 0.5
      ignore: ["lm_head"]
```

### Key Changes:
- **Auto-inference of `mappings`:** The `SmoothQuantModifier` now automatically detects and applies appropriate layer mappings based on the model's architecture, making the modifier more user-friendly and reducing the risk of manual configuration errors.
- **Manual `mappings` parameter removal:** The `mappings` parameter is no longer required in the configuration, as it is inferred dynamically.
- **Backward Compatibility:** Existing configurations that manually specify `mappings` will still be supported, ensuring smooth transition and compatibility with older setups.

### Motivation:
These changes improve usability by automating configuration setup and reducing user overhead, as outlined in the design document: [Link to Design Doc](https://docs.google.com/document/d/1NhZvVmTV9IQ6NqyJM1r3QRMf5MwpBB_Ry_qodykW51U/edit). This also ensures that the quantization recipes are adaptable to various model architectures without manual intervention.

The autoinference of mappings were tested using a Mixtral model: `Isotonic/TinyMixtral-4x248M-MoE`

### CODE REVIEW

File: src/llmcompressor/modifiers/smoothquant/base.py [hunk 1]
```diff
@@ -162,6 +162,17 @@ def on_finalize(self, state: State, **kwargs) -> bool:
 
         return True
 
+    def _infer_mappings_from_model(
+        self, model: Module, mappings: Optional[List[Tuple]] = None
+    ) -> List[Tuple]:
+        if mappings is not None:
+            return mappings
+
+        logger.info("No mappings provided, inferring from model...")
```
Comment (mgoin) [inline src/llmcompressor/modifiers/smoothquant/base.py:171]:
```suggestion
        logger.info("No SmoothQuantModifier.mappings provided, inferring from model...")
```

File: src/llmcompressor/modifiers/smoothquant/utils.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+def get_layer_mappings_from_architecture(architecture: str) -> List[LayerMap]:
+    """
+    :param architecture: str: The architecture of the model
+    :return: list: The layer mappings for the given architecture
+    """
+
+    if architecture not in MAPPINGS_REGISTRY:
+        logger.info(
+            f"Architecture {architecture} not found in mappings. Using default mappings"
```
Comment (mgoin) [inline src/llmcompressor/modifiers/smoothquant/utils.py:54]: Can you add the DEFAULT_SMOOTHQUANT_MAPPINGS to the log please?

Comment (rahul-tuli) [reply src/llmcompressor/modifiers/smoothquant/utils.py]: Done!

File: src/llmcompressor/modifiers/smoothquant/base.py (hunk 1 above)
Comment (rahul-tuli) [reply src/llmcompressor/modifiers/smoothquant/base.py]: Done!

File: src/llmcompressor/modifiers/smoothquant/base.py [hunk 3]
```diff
@@ -124,6 +123,7 @@ def on_initialize(self, state: State, **kwargs) -> bool:
             )
 
         self.ignore = [] if not self.ignore else self.ignore
+        self.mappings = self._infer_mappings_from_model(state.model, self.mappings)
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/smoothquant/base.py:126]: nit: You could make the function name a little bit more precise
```
if mappings is not None:
    self.mappings = self._infer_mappings_from_model(state.model)
```

**Review by kylesayrs — DISMISSED** [review]

**Comment by kylesayrs** [comment]
Should consider adding a sentence like "mappings will normally be automatically inferred, but here's how to create your own custom ones" to https://github.com/vllm-project/llm-compressor/pull/115/files

**Review by kylesayrs — DISMISSED** [review]
Good stuff. As mentioned, including this in the documentation will ensure that this feature actually gets used by users

**Review by kylesayrs — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Pull request merged**